Handle map/unmap events on Wayland's wxGLCanvasEGL

joanbm opened the pull request:
Fixes hiding a wxGLCanvas on Wayland, either directly (`->Show(false)`) or indirectly (e.g. when it is contained in a wxNotebook).

On Wayland, unlike on X11, the window and surface that we create with `wl_egl_window_create` and `eglCreatePlatformWindowSurface` is detached from the GTK widget associated to the canvas, thus it is not automatically mapped or unmapped when the associated GTK widget is. Rather, we need to manually synchronize it with the widget's state.

As Wayland/EGL does not provide any way to temporally hide the window, we need to fully destroy it on unmap, and re-create it on map.

Fixes #22580

---

Annex: Test case that exercises some tricky cases: The necessity of the `map-event` signal instead of `map` when hiding and re-showing toplevel windows containing canvases, and correct handling of re-parenting of canvases between windows.

```c++
#include <wx/app.h>
#include <wx/glcanvas.h>
#include <wx/dcclient.h>
#include <wx/frame.h>
#include <wx/button.h>
#include <wx/sizer.h>

class MyWxApp : public wxApp
{
	wxGLContext *context;

public:
	bool OnInit() override {
		auto window = new wxFrame(nullptr, -1, wxT("Test"));

		auto subWindow = new wxFrame(nullptr, -1, wxT("Sub1"));
		subWindow->Bind(wxEVT_CLOSE_WINDOW, [subWindow](wxCloseEvent &evt) {
			subWindow->Show(false);
		});

		auto subWindow2 = new wxFrame(nullptr, -1, wxT("Sub2"));
		subWindow2->Bind(wxEVT_CLOSE_WINDOW, [subWindow2](wxCloseEvent &evt) {
			subWindow2->Show(false);
		});

		auto canvas = new wxGLCanvas(subWindow);
		context = new wxGLContext(canvas);
		canvas->Bind(wxEVT_PAINT, [this, canvas](wxPaintEvent &evt) {
			context->SetCurrent(*canvas);
			wxPaintDC dc(canvas);

			glClearColor(1.0f, 0.0f, 0.0f, 1.0f);
			glClear(GL_COLOR_BUFFER_BIT);

			canvas->SwapBuffers();
		});

		auto buttonToggleWindows = new wxButton(window, wxID_ANY, wxT("Toggle windows"));
		buttonToggleWindows->Bind(wxEVT_BUTTON, [subWindow, subWindow2](wxCommandEvent &evt) {
			subWindow->Show(!subWindow->IsShown());
			subWindow2->Show(!subWindow2->IsShown());
		});

		auto buttonToggleCanvas = new wxButton(window, wxID_ANY, wxT("Toggle canvas"));
		buttonToggleCanvas->Bind(wxEVT_BUTTON, [canvas](wxCommandEvent &evt) {
			canvas->Show(!canvas->IsShown());
		});

		auto buttonReparent = new wxButton(window, wxID_ANY, wxT("Reparent"));
		buttonReparent->Bind(wxEVT_BUTTON, [subWindow, subWindow2, canvas](wxCommandEvent &evt) {
			canvas->Reparent(canvas->GetParent() == subWindow ? subWindow2 : subWindow);
		});

		auto sizer = new wxBoxSizer(wxHORIZONTAL);
		sizer->Add(buttonToggleWindows, 1, wxEXPAND);
		sizer->Add(buttonToggleCanvas, 1, wxEXPAND);
		sizer->Add(buttonReparent, 1, wxEXPAND);
		window->SetSizer(sizer);
		window->Layout();

		window->Bind(wxEVT_CLOSE_WINDOW, [window, subWindow, subWindow2](wxCloseEvent &evt) {
			subWindow->Destroy();
			subWindow2->Destroy();
			window->Destroy();
		});

		window->Show(true);
		SetTopWindow(window);
		return true;
	}

	int OnExit()
	{
		delete context;
		return 0;
	}
};

IMPLEMENT_APP(MyWxApp)
```

joanbm committed:
Handle map/unmap events on Wayland's wxGLCanvasEGL

Fixes hiding a wxGLCanvas on Wayland, either directly (`->Show(false)`)
or indirectly (e.g. when it is contained in a wxNotebook).

On Wayland, unlike on X11, to show the canvas on the screen, we need to
create a Wayland subsurface. This subsurface is detached from the GTK
widget associated to the canvas, thus it is not automatically mapped or
unmapped when the associated GTK widget is.
Rather, we need to manually keep it in sync with the widget's state.

Knowing what has to be done to map and unmap the canvas, while dealing
with edge cases properly, is not easy to someone not used to Wayland.
When the canvas is mapped, we have this graph of resources:

  EGL Surface (m_surface)
             |
             |
             v
  wl_egl window (m_wlEGLWindow)
             |
             |
             v
     Canvas wl_surface         GDK's toplevel window wl_surface
       (m_wlSurface)        (gdk_wayland_window_get_wl_surface(w))
              \                   which is shown to the user
               \                         ^
                \                       /
                 \                     /
                  v                   /
          Subsurface (m_wlSubsurface) to overlay
          the canvas onto the toplevel window

A simple way would be to destroy everything (m_surface, m_wlEGLWindow,
m_wlSurface, m_wlSubsurface) on unmap, and re-create it again on map.
Inefficiencies aside, this mostly works. However, it can mess with the
current OpenGL context. For example, suppose we have a (fragile)
program that places a canvas inside one of wxNotebook's tabs, and makes
the OpenGL context current only once at startup (e.g. on wxEVT_SHOW).
Switching between tabs will destroy and re-create the EGL Surface, so
the canvas will not be properly rendered when going back to its tab.

So we need to be smarter, and find some way to hide the subsurface
instead. The obvious way would be to unmap the canvas wl_surface
(m_wlSurface), as according to the Wayland spec., "a sub-surface is
hidden [...] if a NULL wl_buffer is applied [to the canvas surface]."
(https://wayland.freedesktop.org/docs/html/apa.html#protocol-spec-wl_subsurface)
However, as far as I can tell, this can't be done. There's no API to
hide an wl_egl window, and directly applying a NULL wl_buffer to the
canvas surface initially hides it, but seems to breaks the associated
window so that it crashes when one attempts to show it again.

So what remains, is destroying the overlay subsurface (m_wlSubsurface).
When doing it, to the spec, "the wl_surface is unmapped immediately.".

And not only does this work, but it also deals with the annoying fact
that on current GTK3 versions, when the GDK's toplevel window is
unmapped, its wl_surface is not just unmapped, but entirely destroyed.
(Side note: This may not have been intended, and has been changed for
GTK4, see: https://gitlab.gnome.org/GNOME/gtk/-/commit/5d3cec5441ca)
So we'd have to re-create the subsurface because of this anyway.

So, this works nicely, and as far as I can tell (documentation is a bit
scarce), there's no problem to leaving m_wlSurface (and its associated
m_wlEGLWindow and m_surface) unmapped in this way.

Fixes #22580

vadz commented:
Thanks a lot once again for fixing this!

I will merge this into master and backport to 3.2 soon if there are no objections.

cc @paulcor @swt2c

joanbm commented:
Grr, I was running some final tests and there's still one test case where it doesn't work (the canvas appears empty), the original `wxNotebook` one: https://github.com/wxWidgets/wxWidgets/issues/22580#issue-1287297419

The problem is that that test case binds the OpenGL context on `wxEVT_SHOW` instead of inside `wxEVT_PAINT` (which I think ought to work):

```c++
    Bind(wxEVT_SHOW, [this](wxShowEvent &) {
        m_renderer->set_active(); // this makes the OpenGL context current

        m_canvas->Bind(wxEVT_PAINT, [this](wxPaintEvent &){
            // ...
        });
    });
```

But the problem is that we wait for GTK's `map-event` signal to create the EGL surface, but `wxEVT_SHOW` happens before it (the order is: 1. `map`, 2. `wxEVT_SHOW`, 3. `map-event`), so at the time it tries to make the OpenGL context current, we haven't yet created the EGL surface. Maybe we should use `map` instead of `map-event`, but this breaks another test case due to the way GDK creates the Wayland surface.

I'll take a look to see what can be done.

vadz commented:
I'm not totally sure about the difference between `map` and `map-event`, but could we, perhaps, generate `wxEVT_SHOW` from the latter instead of the former?

vadz commented:
Looking at [the docs](https://docs.gtk.org/gtk3/signal.Widget.map.html), it seems like `map-event` would be more appropriate in some cases, but not all :-(

We could add `wxEVT_SHOWING` (or `wxEVT_WILL_SHOW`) and make existing `wxEVT_SHOW` synonym for `wxEVT_SHOWN` or vice versa but neither is going to be 100% compatible, of course.

joanbm commented:
Yeah, I suspect that changing `wxEVT_SHOW` in GTK+ to happen after `map-event` is going to lead to all sorts of breakage, since the `map-event` is not emitted synchronous, but rather put in the event loop, so there's a chance for other events to sneak in.

And it's not a hard requirement that we create the EGL window/surface on `map-event` instead of `map`. It's that if you try to do it, you run into a *very* inconvenient GTK/GDK behaviour that makes it hell to re-create the EGL window/surface when the toplevel window containing the canvas is hidden, then re-shown.

For now, I've updated the PR to use `map` again, and a second commit entirely to deal with said GTK/GDK behaviour. However, I think the right solution is to try to get this GTK4 commit: [wayland: Don't destroy the wl_surface on hide()](https://gitlab.gnome.org/GNOME/gtk/-/commit/5d3cec5441ca1fa28e5e77c0572e629eded419c9) backported to GTK3. I'll try to do it, if this can be done, the second commit can be dropped entirely.

With this, all previous tests work, *and* the [wxNotebook sample](https://github.com/wxWidgets/wxWidgets/issues/22580#issue-1287297419) which makes the OpenGL context current on `wxEVT_SHOW` now works on launch.

What *still* doesn't work in that sample, is that if you change to the other tab, and then back to the canvas tab, it no longer render correctly. This is because when changing between tabs, we destroy and re-create the EGL window/surface as needed ([as there is apparently no way to just hide it?](https://stackoverflow.com/q/73941692)). So the context that was bound on `wxEVT_SHOW` becomes stale then. Making the context current on `wxEVT_PAINT` fixes it, as expected.

PS: I also fixed some edge cases I found related to reparenting a wxCanvas to another window.

vadz commented:
Thanks a lot for all your work on this, I can imagine that dealing with all these "oddities" must have been pretty frustrating!

I'm not sure what do you mean by backporting this commit to GTK3, do you really mean doing it in GTK itself? To the best of my knowledge, GTK3 doesn't get any updates any longer...

Anyhow, I'm more than ready to fix this if it fixes all the issues (I don't doubt that it does, but I just haven't had time to check them myself yet) even in the current form. The code is more complicated than ideal, but actually making things work for the users is definitely more important than code elegance in this case.

Thanks again!

joanbm commented:
> Thanks a lot for all your work on this, I can imagine that dealing with all these "oddities" must have been pretty frustrating!

Thanks! Indeed.

> I'm not sure what do you mean by backporting this commit to GTK3, do you really mean doing it in GTK itself? To the best of my knowledge, GTK3 doesn't get any updates any longer...

Yes, in GTK itself. I don't think it gets any new features but it [definitely gets some bug fixes](https://gitlab.gnome.org/GNOME/gtk/-/commits/gtk-3-24/). So it *may* be possible to get it in...

---

Finally, I may also have figured out the answer to your StackOverflow question: To hide it, you can just destroy the subsurface (i.e. `m_wlSubsurface` in our code) that is associated to its surface when the GTK widget unmaps, and the EGL surface is immediately gone from the screen. You don't need to destroy the EGL window or surface.

I'd swear that like you, I tried to do this before, and it didn't work. But I was probably doing something wrong. (And unfortunately with those OpenGL things, you're lucky if you get a crash instead of some weird deadlock, graphical glitch...)

I have yet to fully test and review this (I'll update this PR), but with it, the wxNotebook example should fully work on Wayland, including changing tabs, without any code changes.

joanbm commented:
PR updated, I believe that it's the right approach now.

After understanding that you *just* need to destroy/create the Wayland subsurface to unmap/map the canvas, and not the entire EGL surface and wl_egl window, I finally got the [wxNotebook reproducer](https://github.com/wxWidgets/wxWidgets/issues/22580#issue-1287297419) that binds the OpenGL context only once on `wxEVT_SHOW` to work when switching tabs.

While at the some time, I could go back to the much simpler solution of using `map-event` to ensure that GDK has created the Wayland surface associated to the toplevel window, instead of needing either a backport of [wayland: Don't destroy the wl_surface on hide()](https://gitlab.gnome.org/GNOME/gtk/-/commit/5d3cec5441ca1fa28e5e77c0572e629eded419c9) to GTK3 or quite some ugly code.

I also ran various combinations of programs (wxNotebook reproducer/show-hide-reparent reproducer/SLADE/KiCad) on various compositors (Weston/Sway/GNOME-Shell) and on various OSs (Debian Bullseye VM / Arch Linux baremetal) and could find no problems related to the change.

The only noteworthy testing find is that when hiding then quickly re-showing a toplevel window containing a canvas on GNOME-Shell on my Arch Linux install, it deadlocks inside eglSwapBuffers. But aside from being a minor issue, it's think it's an unrelated issue, and this wasn't working before anyway.

vadz commented on src/unix/glegl.cpp:
```diff
@@ -494,21 +519,25 @@ bool wxGLCanvasEGL::CreateSurface()
+        // widget's underlying Wayland surface has been created.
+        // Otherwise, gdk_wayland_window_get_wl_surface may return nullptr,
+        // for example when hiding then showing a window containing a canvas.
+        gtk_widget_add_events(m_widget, GDK_STRUCTURE_MASK);
```
Interesting, I was a bit surprised we needed this because I knew we were already getting `map-event` for wxTLW without doing it, but it looks like GTK enables this event mask automatically for TLWs only.

vadz reviewed (approved):
Very nice, thanks a lot! Not only it works now, but the code is also pretty clear (separate thanks for all the comments) and straightforward.

If there are no objections I'm going to merge (and backport to 3.2) this in a couple of days.

Thanks again!

swt2c reviewed (approved):
@joanbm thanks a lot for your work on this!  The changes look good to me.

swt2c commented:
> The only noteworthy testing find is that when hiding then very quickly re-showing a toplevel window containing a canvas on GNOME-Shell on my Arch Linux install, it deadlocks inside eglSwapBuffers. But aside from being a minor issue, it's think it's an unrelated issue, and this wasn't working before anyway.

That's strange about the deadlock issue that you saw.  Using the frame callback handler to tell us when it's OK to eglSwapBuffers seems like it ought to take care of that issue (at least that's why I started using it in the first place).

joanbm commented:
> > The only noteworthy testing find is that when hiding then very quickly re-showing a toplevel window containing a canvas on GNOME-Shell on my Arch Linux install, it deadlocks inside eglSwapBuffers. But aside from being a minor issue, it's think it's an unrelated issue, and this wasn't working before anyway.
> 
> That's strange about the deadlock issue that you saw. Using the frame callback handler to tell us when it's OK to eglSwapBuffers seems like it ought to take care of that issue (at least that's why I started using it in the first place).

Yeah, I took a quick look and I didn’t see anything immediately wrong with the frame callbacks, but it just deadlocks. It’s definitely on my todo list to look at this (though no promises). I just hope at the end of the day it’s something dumb instead of a rabbit hole, I’ve had enough of those recently already :)

imciner2 commented:
> I also ran various combinations of programs (wxNotebook reproducer/show-hide-reparent reproducer/SLADE/KiCad) on various compositors (Weston/Sway/GNOME-Shell) and on various OSs (Debian Bullseye VM / Arch Linux baremetal) and could find no problems related to the change.

Thanks for testing this (especially with KiCad), since one of my worries was if destroying/creating the canvas would lead to issues with existing programs (I was concerned that maybe there would be attempts to draw on it once it was destroyed).

> The only noteworthy testing find is that when hiding then very quickly re-showing a toplevel window containing a canvas on GNOME-Shell on my Arch Linux install, it deadlocks inside eglSwapBuffers.

Does this happen on all programs, or did you just observe it on specific ones you tested?

vadz commented:
I don't know if it's worth waiting for more reviews, so let's merge this and see if anybody finds any new problems.

Thanks again Joan for all your work on this!

joanbm commented:
> > > The only noteworthy testing find is that when hiding then very quickly re-showing a toplevel window containing a canvas on GNOME-Shell on my Arch Linux install, it deadlocks inside eglSwapBuffers. But aside from being a minor issue, it's think it's an unrelated issue, and this wasn't working before anyway.
> > 
> > 
> > That's strange about the deadlock issue that you saw. Using the frame callback handler to tell us when it's OK to eglSwapBuffers seems like it ought to take care of that issue (at least that's why I started using it in the first place).
> 
> Yeah, I took a quick look and I didn’t see anything immediately wrong with the frame callbacks, but it just deadlocks. It may have something to do with the window open/close animations (Sway doesn’t have those), as the timing for how quickly how have to click for the deadlock to happen is suspiciously close to the animation length. It’s definitely on my todo list to look at this (though no promises). I just hope at the end of the day it’s something dumb instead of a rabbit hole, I’ve had enough of those recently already :)

FWIW this does not happen anymore now that `eglSwapInterval(display, 0)` from #23909 / #24014 is merged (which makes sense since it disables the logic that can make `eglSwapBuffers()` block), though I'm not sure what the actual root cause is/was.
